### PR TITLE
fix(security): #CO-1774 add several checks on back api for displaying ticket content called event

### PR DESCRIPTION
**Pull request by Gregprod33**
## Describe your changes
This correction was especially made to secure @Get("/events/:id") API in postman calls. 
Before displaying the ticket, we now check whether the user is SUPER_ADMIN or LOCAL_ADMIN.
If he's SUPER_ADMIN, he can see all the tickets. If he's ADMIN_LOCAL, he can only see the tickets belonging to his structure.
Otherwise, he can only see his own tickets.

To do that, the method listEvents in TicketService was replaced to become getListEvents which now return a future.
Moreover, when this method is called, a new helper TicketEventHelper comes with its method shouldRenderEvent to proceed all the checks.

Dependences added for this project :
 testCompile "org.powermock:powermock-api-mockito2:$powerMockVersion"
 testCompile "org.powermock:powermock-module-junit4:$powerMockVersion"
 testCompile "org.powermock:powermock-module-testng:$powerMockVersion"
 
 In order to use in Unit tests 

## Checklist tests
On postman, log with super admin account and try to access to every created ticket.
With ADMIN_LOCAL account, try to access tickets from the structure you belong to and other structure, which should lead to a 401 code error.
At least, try a random account with no specific function and try to access your tickets or other tickets.
## Issue ticket number and link
[CO-1774](https://jira.support-ent.fr/browse/CO-1774)
## Checklist before requesting a review (magic string, indentation, comment/documentation...)

- [x] I have detailed the tests to do in my feature/fix in order to prevent consequents regressions (must specify in **Checklist tests**)
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (API Doc etc...) - (must specify in **Description** for target version)
- [ ] If it is a consequent feature, I have added thorough tests.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] Any dependent changes have been added to this project (must specify in **Description**)

### CODE REVIEW

File: src/main/java/net/atos/entng/support/helpers/impl/TicketEventHelperImpl.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+
+    /**
+     * @param user        : user
+     * @param functions   : aim to check if user is an admin
+     * @param eventResult : the result of handler passed in listEvents method
+     */
+    @Override
+    public boolean shouldRenderEvent(UserInfos user, Map<String, UserInfos.Function> functions, JsonArray eventResult) {
+        String owner = eventResult.getJsonObject(0).getString("user_id");
```
Comment (flmartineau) [inline src/main/java/net/atos/entng/support/helpers/impl/TicketEventHelperImpl.java:20]: Magic strings

File: src/main/java/net/atos/entng/support/helpers/impl/TicketEventHelperImpl.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        String schoolId = eventResult.getJsonObject(0).containsKey("school_id") ? eventResult.getJsonObject(0).getString("school_id") : null;
```
Comment (flmartineau) [inline src/main/java/net/atos/entng/support/helpers/impl/TicketEventHelperImpl.java:21]: Ajoute une vérif sur eventResult (eventResult !== null && !eventResult.isEmpty)

**Review by flmartineau — APPROVED** [review]

File: src/main/java/net/atos/entng/support/helpers/impl/TicketEventHelperImpl.java (hunk 1 above)
Comment (Gregprod33) [reply src/main/java/net/atos/entng/support/helpers/impl/TicketEventHelperImpl.java]: Modifié :)

File: src/main/java/net/atos/entng/support/helpers/impl/TicketEventHelperImpl.java (hunk 2 above)
Comment (Gregprod33) [reply src/main/java/net/atos/entng/support/helpers/impl/TicketEventHelperImpl.java]: Modifié

File: src/main/java/net/atos/entng/support/controllers/TicketController.java [hunk 3]
```diff
@@ -889,19 +891,33 @@ public void createTicketHistoMultiple(List<Integer> idList, String event, Intege
         }
     }
 
+
     @Get("/events/:id")
     @ApiDoc("Get historization of a ticket")
     public void getHistorization(final HttpServerRequest request) {
```
Comment (BastienBenard) [inline src/main/java/net/atos/entng/support/controllers/TicketController.java]: Pas de droit ? Il faudrait rajouter un AUTHENTICATED

File: src/main/java/net/atos/entng/support/helpers/TicketEventHelper.java [hunk 4]
```diff
@@ -0,0 +1,10 @@
+package net.atos.entng.support.helpers;
+
+import io.vertx.core.json.JsonArray;
+import org.entcore.common.user.UserInfos;
+
+import java.util.Map;
+
+public interface TicketEventHelper {
+    public boolean shouldRenderEvent(UserInfos user, Map<String, UserInfos.Function> functions, JsonArray eventResult);
+}
```
Comment (BastienBenard) [inline src/main/java/net/atos/entng/support/helpers/TicketEventHelper.java:10]: Il n'y a pas besoin d'interface pour les helpers, puis que tout les méthode sont sensé être statics

File: src/main/java/net/atos/entng/support/controllers/TicketController.java (hunk 3 above)
Comment (Gregprod33) [reply src/main/java/net/atos/entng/support/controllers/TicketController.java]: C'est fait

File: src/main/java/net/atos/entng/support/helpers/IModelHelper.java
Comment (Ozen40) [inline src/main/java/net/atos/entng/support/helpers/IModelHelper.java:1]: [IModelHelper](https://github.com/OPEN-ENT-NG/crre/blob/1770e7e367ee2888e1b930472cc6c4a527597555/src/main/java/fr/openent/crre/helpers/IModelHelper.java) provient de CRRE

File: src/main/java/net/atos/entng/support/helpers/RequestHelper.java
Comment (Ozen40) [inline src/main/java/net/atos/entng/support/helpers/RequestHelper.java:1]: [RequestHelper](https://github.com/OPEN-ENT-NG/minibadge/blob/dev/src/main/java/fr/openent/minibadge/helper/RequestHelper.java#L44) provient de Minibadge

File: src/main/java/net/atos/entng/support/helpers/StringHelper.java
Comment (Ozen40) [inline src/main/java/net/atos/entng/support/helpers/StringHelper.java:1]: [StringHelper](https://github.com/OPEN-ENT-NG/crre/blob/1770e7e367ee2888e1b930472cc6c4a527597555/src/main/java/fr/openent/crre/helpers/StringHelper.java) provient de CRRE

File: src/main/java/net/atos/entng/support/model/IModel.java
Comment (Ozen40) [inline src/main/java/net/atos/entng/support/model/IModel.java:1]: [IModel](https://github.com/OPEN-ENT-NG/crre/blob/1770e7e367ee2888e1b930472cc6c4a527597555/src/main/java/fr/openent/crre/model/IModel.java) provient de CRRE

File: src/main/java/net/atos/entng/support/controllers/TicketController.java [hunk 5]
```diff
@@ -834,22 +836,26 @@ public void createTicketHistoMultiple(List<Integer> idList, String event, Intege
...
+    @ResourceFilter(OwnerOrLocalAdmin.class)
     public void getHistorization(final HttpServerRequest request) {
         final String ticketId = request.params().get("id");
         UserUtils.getUserInfos(eb, request, user -> {
             if (user != null) {
-                ticketServiceSql.listEvents(ticketId, arrayResponseHandler(request));
+                ticketServiceSql.getlistEvents(ticketId)
+                        .onSuccess(result -> renderJson(request, RequestHelper.addAllValue(new JsonObject(), result).getJsonArray(Ticket.ALL)))
+                        .onFailure(err -> log.error(String.format("[Support@%s::getListEvent] Failed to display events of ticket",
+                                this.getClass().getSimpleName())));
```
Comment (qpic-cgi) [inline src/main/java/net/atos/entng/support/controllers/TicketController.java:851]: Dans ton exception, tu ne log pas le détail de ton erreur. 
Ensuite, ici, si tu as une as une erreur, ton front va rester en pending sans jamais savoir que ton back a en fait fail.

File: src/main/java/net/atos/entng/support/helpers/FutureHelper.java [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    public static Handler<Either<String, JsonArray>> handlerJsonArray(Promise<JsonArray> promise) {
+        return event -> {
+            if (event.isRight()) {
+                promise.complete(event.right().getValue());
+            } else {
+                LOGGER.error(event.left().getValue());
+                promise.fail(event.left().getValue());
+            }
+        };
+    }
```
Comment (qpic-cgi) [inline src/main/java/net/atos/entng/support/helpers/FutureHelper.java:44]: Evite de ramener du code déprécier dans ton module ^^

**Review by qpic-cgi — CHANGES_REQUESTED** [review]

File: src/main/java/net/atos/entng/support/helpers/FutureHelper.java (hunk 6 above)
Comment (Ozen40) [reply src/main/java/net/atos/entng/support/helpers/FutureHelper.java]: J'ai supprimé ce fichier ça il était utilisé avec l'ancienne méthode mais plus maintenant (en plus il faisait la même chose que mon PromiseHelper)

**Review by qpic-cgi — APPROVED** [review]

**Pull request merged**